File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.cc
/*
아이디어 : 1. (1 -> v1 -> v2 -> n) VS (1 -> v2 -> v1 -> n) 비교하기
알고리즘 : dijkstra(graph, dist_table, INF)
자료구조 : pq 
*/

#include <bits/stdc++.h>

#define MAX 801
#define INF 1e7

using namespace std;

int n, e;
vector<pair<int, int> > graph[MAX];
int v1, v2;

int dijkstra(int start, int end) {
    // 거리 테이블 정의
    int dist_table[MAX];
    fill(dist_table, dist_table + MAX, INF);
    
    priority_queue<pair<int, int> > pq;
    pq.push({0, start}); // {거리, 노드} 순으로 저장
    dist_table[start] = 0;
    
    // pq가 빌 때까지
    while(!pq.empty()) {
        int dist = -pq.top().first;
        int now = pq.top().second;
        pq.pop();
        
        // 이미 처리된 적이 있다면
        if (dist_table[now] < dist) continue;
        
        // 해당 노드와 인접한 노드들 확인
        for (int i = 0; i < graph[now].size(); i++) {
            int v = graph[now][i].first;
            int d = graph[now][i].second;
            
            int cost = dist + d;
            
            // 갱신 거리가 기존 거리보다 짧다면
            if (cost < dist_table[v]) {
                dist_table[v] = cost;
                pq.push({-cost, v});
            }
        }
        
    }
    return dist_table[end];
}

int main(void) {
    ios::sync_with_stdio(false);
    cin.tie(NULL); cout.tie(NULL);
    
    cin >> n >> e;
    
    // 간선 정보 입력
    for(int i = 0; i < e; i++) {
        int a, b, c;
        cin >> a >> b >> c;
        // 양방향
        graph[a].push_back({b, c});
        graph[b].push_back({a, c});
    }
    
    cin >> v1 >> v2;
    
    // 1 -> v1 -> v2 -> n
    int value1 = dijkstra(1, v1) + dijkstra(v1, v2) + dijkstra(v2, n);
    int value2 = dijkstra(1, v2) + dijkstra(v2, v1) + dijkstra(v1, n);
    
    int result = min(value1, value2);
    
    if (result < INF)
        cout << result << '\n';
    else 
        cout << -1 << '\n';
    
    return 0;
}

File: 백준/Gold/1504. 특정한 최단 경로/특정한 최단 경로.py
# 아이디어 : 1. graph, distance, heap 정의
#            2. 양방향성
#            3. (1 -> u -> v -> n) or (1 -> v -> u -> n) 둘 중에 하나가 최소
# 알고리즘 : dijkstra
# 자료구조 : heapq

import heapq
import sys
input = sys.stdin.readline
INF = int(1e9)

n, m = map(int, input().split())

# 그래프 정의
graph = [[] for _ in range(n + 1)]
for _ in range(m):
    a, b, c = map(int, input().split())
    graph[a].append((b, c))
    graph[b].append((a, c))
    


u, v = map(int, input().split())

def dijkstra(start):
    # 거리 테이블 정의
    distance = [INF for _ in range(n + 1)]
    
    # 힙큐 및 시작노드 정의
    hq = []
    heapq.heappush(hq, (0, start)) # (거리, 노드)순으로 저장
    distance[start] = 0
    
    # 힙큐가 빌 때까지
    while hq:
        dist, now = heapq.heappop(hq)
        
        # 이미 처리된 적이 있다면
        if distance[now] < dist:
            continue
            
        for v, d in graph[now]:
            cost = dist + d
            # 기존 거리보다 갱신 거리가 크다면
            if distance[v] > cost:
                distance[v] = cost
                heapq.heappush(hq, (cost, v))
                
    return distance

dist_1 = dijkstra(1)
dist_u = dijkstra(u)
dist_v = dijkstra(v)
                
path1 = dist_1[u] + dist_u[v] + dist_v[n]
path2 = dist_1[v] + dist_v[u] + dist_u[n]

result = min(path1, path2)

if result >= INF:
    print(-1)
else :
    print(result)